server/src/internalClusterTest/java/org/elasticsearch/snapshots/AbortedRestoreIT.java

```diff
@@ -0,0 +1,352 @@
+            assertThat(blockRepository.streams().filter(BlockingInputStream::isBlocked).count(), equalTo((long) activeSnapshotThreads));
+        }, 30L, TimeUnit.SECONDS);
+
+        logger.info("--> aborting restore by deleting the index");
+        assertAcked(client().admin().indices().prepareDelete(indexName));
+
+        // Total number of blobs that have been opened from the blob store
+        final long totalBlobsRead = blockRepository.streams().count();
+
+        // Total number of bytes that have been read from the blob store
+        final long totalBytesRead = blockRepository.streams().mapToLong(BlockingInputStream::getCount).sum();
+
+        logger.info("--> unblocking node [{}]", dataNode);
+        blockRepository.unblock();
+        assertThat(blockRepository.streams().noneMatch(BlockingInputStream::isBlocked), is(true));
+
+        logger.info("--> restore should have failed");
+        final RestoreSnapshotResponse restoreSnapshotResponse = future.get();
+        assertThat(restoreSnapshotResponse.getRestoreInfo().failedShards(), equalTo(numPrimaries));
+        assertThat(restoreSnapshotResponse.getRestoreInfo().successfulShards(), equalTo(0));
+
+        logger.info("--> waiting for snapshot thread pool to be empty");
+        assertBusy(() -> {
+            ThreadPool threadPool = internalCluster().getInstance(ClusterService.class, dataNode).getClusterApplierService().threadPool();
+            int activeSnapshotThreads = -1;
+            for (ThreadPoolStats.Stats threadPoolStats : threadPool.stats()) {
+                if (threadPoolStats.getName().equals(ThreadPool.Names.SNAPSHOT)) {
+                    activeSnapshotThreads = threadPoolStats.getActive();
+                    break;
+                }
+            }
+            assertThat(activeSnapshotThreads, equalTo(0));
+            assertThat(blockRepository.streams().filter(BlockingInputStream::isBlocked).count(), equalTo(0L));
+        }, 30L, TimeUnit.SECONDS);
+
+        assertThat("No more blobs should have been opened from the blob store",
+            blockRepository.streams().count(), equalTo(totalBlobsRead));
+        assertThat("No more bytes should have been read from the blob store",
+            blockRepository.streams().mapToLong(BlockingInputStream::getCount).sum(), equalTo(totalBytesRead));
+    }
+
+    /**
+     * A blob store repository that blocks read operations on {@link InputStream} when the {@code blockStreams} flag
+     * is set to true. It also keep track of the number of bytes read from {@link InputStream} it opens.
+     */
+    public static class BlockingDataFileReadsRepository extends FsRepository {
+
+        static final String TYPE = "block_on_data_file_reads";
+
+        private final List<BlockingInputStream> streams = Collections.synchronizedList(new ArrayList<>());
+        private volatile boolean blockStreams = false;
+
+        public BlockingDataFileReadsRepository(
```

Comment:
Maybe just override `org.elasticsearch.snapshots.mockstore.MockRepository#blockOnDataFiles` and add a method to verify no more interactions (could just add a flag that throws an `AssertionError` past a certain point)? Then you can just wait for the snapshot threads to all block, unblock and wait for the snapshot pool to be empty. Much less new test code to maintain and tests the same thing pretty much?

Reply:
I agree it would be less code to maintain, but `MockRepository` blocks before reading any bytes and I wanted something that allows to read few bytes then block. But now I'm thinking to it again it does not make more difference as `InputStream` opened by FS blob containers do not have any abort method. I'm going to take a look again at using `MockRepository`.